api/v1alpha3/virtualmachine_types.go

```diff
@@ -208,28 +208,53 @@ const (
 	VirtualMachinePowerOpModeTrySoft VirtualMachinePowerOpMode = "TrySoft"
 )
 
+type VirtualMachineImageRef struct {
+	// Kind describes the type of image, either a namespace-scoped
+	// VirtualMachineImage or cluster-scoped ClusterVirtualMachineImage.
+	Kind string `json:"kind"`
+
+	// Name refers to the name of a VirtualMachineImage resource in the same
+	// namespace as this VM or a cluster-scoped ClusterVirtualMachineImage.
+	Name string `json:"name"`
+}
+
 // VirtualMachineSpec defines the desired state of a VirtualMachine.
 type VirtualMachineSpec struct {
+
+	// +optional
+
+	// Image describes the reference to the VirtualMachineImage or
+	// ClusterVirtualMachineImage resource used to deploy this VM.
+	//
+	// Please note, unlike the field spec.imageName, the value of
+	// spec.image.name MUST be a Kubernetes object name.
+	//
+	// Please also note, when this field is non-empty, the value of the
+	// field spec.imageName is ignored.
+	Image *VirtualMachineImageRef `json:"image,omitempty"`
+
+	// +optional
+
 	// ImageName describes the name of the image resource used to deploy this
 	// VM.
 	//
 	// This field may be used to specify the name of a VirtualMachineImage
 	// or ClusterVirtualMachineImage resource. The resolver first checks to see
-	// if there is a VirtualMachineImage with the specified name. If no
-	// such resource exists, the resolver then checks to see if there is a
-	// ClusterVirtualMachineImage resource with the specified name in the same
-	// Namespace as the VM being deployed.
+	// if there is a VirtualMachineImage with the specified name in the
+	// same namespace as the VM being deployed. If no such resource exists, the
+	// resolver then checks to see if there is a ClusterVirtualMachineImage
+	// resource with the specified name.
 	//
 	// This field may also be used to specify the display name (vSphere name) of
 	// a VirtualMachineImage or ClusterVirtualMachineImage resource. If the
 	// display name unambiguously resolves to a distinct VM image (among all
 	// existing VirtualMachineImages in the VM's namespace and all existing
-	// ClusterVirtualMachineImages), then a mutation webhook updates this field
-	// with the VM image resource name. If the display name resolves to multiple
-	// or no VM images, then the mutation webhook denies the request and outputs
-	// an error message accordingly.
+	// ClusterVirtualMachineImages), then a mutation webhook updates the
+	// spec.image field with the reference to the resolved VM image. If the
+	// display name resolves to multiple or no VM images, then the mutation
+	// webhook denies the request and returns an error.
 	//
-	// +optional
+	// Please note, this field is ignored when spec.image is non-empty.
```

Comment:
I'm okay with this. A UX question I had.  

Why 
- the ignore 'ImageName' when 'spec.image' is non-empty strategy,  instead of 
- making 'ImageName' and 'spec.Image' mutually exclusive via a validation error?. 

Wanted to know your thoughts since I'm sure you've considered this and we could have an unused ImageName in spec now.

Reply:
> making 'ImageName' and 'spec.Image' mutually exclusive via a validation error?.

We cannot do that without mutating `spec.imageName` to `empty` once the image is resolved. One purpose of this PR is to stop mutating the field in place.

Reply:
See https://github.com/vmware-tanzu/vm-operator/pull/470#discussion_r1567949461 for other question.